### PR TITLE
extend lifetime of query context if some fragments are pending on wire

### LINKED ISSUE #3734: be crash when pipeline turn on
### Steps to reproduce the behavior (Required)
1. CREATE TABLE '...'
2. INSERT INTO '....'
3. SELECT  '....'

### Real behavior (Required)
```
=================================================================
==7918==ERROR: AddressSanitizer: heap-use-after-free on address 0x613000882eb0 at pc 0x000006d77b1d bp 0x7fd4bd3459d0 sp 0x7fd4bd3459c8
READ of size 8 at 0x613000882eb0 thread T362
    #0 0x6d77b1c in __gnu_cxx::__normal_iterator<std::shared_ptr<starrocks::pipeline::PipelineDriver>*, std::vector<std::shared_ptr<starrocks::pipeline::PipelineDriver>, std::allocator<std::shared_ptr<starrocks::pipeline::PipelineDriver> > > >::__normal_iterator(std::shared_ptr<starrocks::pipeline::PipelineDriver>* const&) /usr/include/c++/10.3.0/bits/stl_iterator.h:976
    #1 0x6d702b2 in std::vector<std::shared_ptr<starrocks::pipeline::PipelineDriver>, std::allocator<std::shared_ptr<starrocks::pipeline::PipelineDriver> > >::begin() /usr/include/c++/10.3.0/bits/stl_vector.h:812
    #2 0x6d6021f in starrocks::pipeline::FragmentExecutor::execute(starrocks::ExecEnv*) /root/starrocks/be/src/exec/pipeline/fragment_executor.cpp:291
    #3 0x55bacdb in starrocks::PInternalServiceImpl<doris::PBackendService>::_exec_plan_fragment(brpc::Controller*) (/home/disk1/dorisdb/pipeline_sqlancer_master_release/be/lib/starrocks_be+0x55bacdb)
    #4 0x55b6223 in starrocks::PInternalServiceImpl<doris::PBackendService>::exec_plan_fragment(google::protobuf::RpcController*, starrocks::PExecPlanFragmentRequest const*, starrocks::PExecPlanFragmentResult*, google::protobuf::Closure*) (/home/disk1/dorisdb/pipeline_sqlancer_master_release/be/lib/starrocks_be+0x55b6223)
    #5 0x5ef03f5 in doris::PBackendService::CallMethod(google::protobuf::MethodDescriptor const*, google::protobuf::RpcController*, google::protobuf::Message const*, google::protobuf::Message*, google::protobuf::Closure*) /root/starrocks/gensrc/build/gen_cpp/[doris_internal_service.pb.cc:201](http://doris_internal_service.pb.cc:201/)
    #6 0x9db4aed in brpc::policy::ProcessRpcRequest(brpc::InputMessageBase*) /var/local/starrocks/thirdparty/src/incubator-brpc-0.9.7/src/brpc/policy/baidu_rpc_protocol.cpp:496
    #7 0x9dab556 in brpc::ProcessInputMessage(void*) /var/local/starrocks/thirdparty/src/incubator-brpc-0.9.7/src/brpc/input_messenger.cpp:135
    #8 0x9dac402 in brpc::RunLastMessage::operator()(brpc::InputMessageBase*) /var/local/starrocks/thirdparty/src/incubator-brpc-0.9.7/src/brpc/input_messenger.cpp:141
    #9 0x9dac402 in std::unique_ptr<brpc::InputMessageBase, brpc::RunLastMessage>::~unique_ptr() /usr/include/c++/10.3.0/bits/unique_ptr.h:361
    #10 0x9dac402 in brpc::InputMessenger::OnNewMessages(brpc::Socket*) /usr/include/c++/10.3.0/bits/unique_ptr.h:355
    #11 0x9e529fd in brpc::Socket::ProcessEvent(void*) /var/local/starrocks/thirdparty/src/incubator-brpc-0.9.7/src/brpc/socket.cpp:1017
    #12 0x9ef762e in bthread::TaskGroup::task_runner(long) /var/local/starrocks/thirdparty/src/incubator-brpc-0.9.7/src/bthread/task_group.cpp:296
    #13 0x9eebba0 in bthread_make_fcontext (/home/disk1/dorisdb/pipeline_sqlancer_master_release/be/lib/starrocks_be+0x9eebba0)

0x613000882eb0 is located 176 bytes inside of 360-byte region [0x613000882e00,0x613000882f68)
freed by thread T362 here:
    #0 0x41695b7 in operator delete(void*, unsigned long) ../../.././libsanitizer/asan/asan_new_delete.cpp:172

previously allocated by thread T362 here:
    #0 0x4168797 in operator new(unsigned long) ../../.././libsanitizer/asan/asan_new_delete.cpp:99

Thread T362 created by T0 here:
    #0 0x4112492 in __interceptor_pthread_create ../../.././libsanitizer/asan/asan_interceptors.cpp:214
    #1 0x9ef32ac in bthread::TaskControl::init(int) /var/local/starrocks/thirdparty/src/incubator-brpc-0.9.7/src/bthread/task_control.cpp:165
    #2 0x9ee86a1 in bthread::get_or_new_task_control() /var/local/starrocks/thirdparty/src/incubator-brpc-0.9.7/src/bthread/bthread.cpp:91
    #3 0x9ee7c8c in bthread::start_from_non_worker(unsigned long*, bthread_attr_t const*, void* (*)(void*), void*) /var/local/starrocks/thirdparty/src/incubator-brpc-0.9.7/src/bthread/bthread.cpp:128
    #4 0x9ee7c8c in bthread_start_background /var/local/starrocks/thirdparty/src/incubator-brpc-0.9.7/src/bthread/bthread.cpp:193
    #5 0x9da491e in GlobalInitializeOrDieImpl /var/local/starrocks/thirdparty/src/incubator-brpc-0.9.7/src/brpc/global.cpp:604
    #6 0x7fd59d6b31ca in __pthread_once_slow (/lib64/libpthread.so.0+0x61ca)

SUMMARY: AddressSanitizer: heap-use-after-free /usr/include/c++/10.3.0/bits/stl_iterator.h:976 in __gnu_cxx::__normal_iterator<std::shared_ptr<starrocks::pipeline::PipelineDriver>*, std::vector<std::shared_ptr<starrocks::pipeline::PipelineDriver>, std::allocator<std::shared_ptr<starrocks::pipeline::PipelineDriver> > > >::__normal_iterator(std::shared_ptr<starrocks::pipeline::PipelineDriver>* const&)
Shadow bytes around the buggy address:
  0x0c2680108580: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2680108590: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801085a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801085b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801085c0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c26801085d0: fd fd fd fd fd fd[fd]fd fd fd fd fd fd fd fd fd
  0x0c26801085e0: fd fd fd fd fd fd fd fd fd fd fd fd fd fa fa fa
  0x0c26801085f0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2680108600: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2680108610: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2680108620: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==7918==ABORTING
```
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
`PIPELINE_SQLANCER_MASTER_RELEASE bf5724e`

### PR DESCRIPTION
**Pull request by satanson**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #
https://github.com/StarRocks/starrocks/issues/3734
## Problem Summary(Required) ：

When FE delivers N fragment instance to BE, BE receives fist n fragment instance, so QueryContext keep n active fragment instance. when n active fragments are finished, then num_active_fragments counts down to zero, but there are still (N-n) fragments are pending on the wire, sometimes, one of the pending fragment instance is received and BE invoke QueryContextManager::instance()->get_or_register() to obtain the existing query context, but at the same moment, the query context is removed since one of pipeline executor thread to see that there are no active fragment instances in query context.

so, in this PR, two improvements are added:
1. a query context can be cleaned safely only if it receives all its fragment instances and all of its fragment instances are termintes, otherwise, the query context is moved to the second chance map to wait for its expiration. at the waiting time, if some   fragments comes, the query context would be put back into the context map. until the query context is expired, then it is removed. and the remove operation and get_or_registor operation are protected by the mutex.
2. when a query context is put into second chance map, extend its lifetime, so it can wait for late-coming fragment instances. later DescriptorTable cache PR will benifits from this mechanism.

### CODE REVIEW

File: be/src/exec/pipeline/query_context.cpp [hunk 1]
```diff
@@ -28,43 +28,116 @@ void QueryContext::cancel(const Status& status) {
     _fragment_mgr->cancel(status);
 }
 
-QueryContextManager::QueryContextManager() = default;
+static constexpr size_t QUERY_CONTEXT_MAP_SLOT_NUM = 64;
+QueryContextManager::QueryContextManager()
+        : _mutexes(QUERY_CONTEXT_MAP_SLOT_NUM),
+          _context_maps(QUERY_CONTEXT_MAP_SLOT_NUM),
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/query_context.cpp:34]: Shard map? Why not a concurrent map?

Comment (mofeiatwork) [reply be/src/exec/pipeline/query_context.cpp:34]: Otherwise the shard map could be extracted to a common class?

Comment (satanson) [reply be/src/exec/pipeline/query_context.cpp:34]: very good suggest, but concurrent map in this PR,  a mutex manage to two map, it's not a general-purpose concurrent map.

**Comment by stdpain** [comment]
run starrocks_be_unittest

**Review by stdpain — DISMISSED** [review]

**Review by stdpain — DISMISSED** [review]

File: be/src/exec/pipeline/query_context.cpp [hunk 2]
```diff
@@ -28,43 +28,121 @@ void QueryContext::cancel(const Status& status) {
...
-    if (iter != _contexts.end()) {
-        iter->second->increment_num_fragments();
-        return iter->second.get();
+    size_t i = std::hash<size_t>()(query_id.lo) & QUERY_CONTEXT_MAP_SLOT_NUM_MASK;
+    auto& mutex = _mutexes[i];
+    auto& context_map = _context_maps[i];
+    auto& sc_map = _second_chance_maps[i];
+
+    {
+        // require read lock
```
Comment (kangkaisen) [inline be/src/exec/pipeline/query_context.cpp:50]: These comments seems needn't

Comment (satanson) [reply be/src/exec/pipeline/query_context.cpp:50]: for context_map,  write operations are very rare, O(num_queries).  but read operations are very often,  O(num_fragment_instances), so use read lock to reduce contest.

File: be/src/exec/pipeline/query_context.h [hunk 3]
```diff
@@ -75,15 +76,17 @@ class QueryContextManager {
...
-    QueryContextPtr remove(const TUniqueId& query_id);
+    void remove(const TUniqueId& query_id);
 
 private:
-    //TODO(by satanson)
-    // A multi-shard map may be more efficient
-    std::mutex _lock;
-    std::unordered_map<TUniqueId, QueryContextPtr> _contexts;
+    std::vector<std::shared_mutex> _mutexes;
+    std::vector<std::unordered_map<TUniqueId, QueryContextPtr>> _context_maps;
```
[the remark this answers is not in the document]
Comment (satanson) [reply be/src/exec/pipeline/query_context.h:88]: It's needless, the performance bottleneck comes from write/read contest.

**Review by stdpain — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**